File: lib/receipt_generator.py
from decimal import Decimal

from lib.item import Item
from lib.receipt import Receipt


class ReceiptGenerator:
    def __init__(self, input_lines):
        self.input = input_lines

    def get_receipt(self):
        items = self.create_items()
        Receipt(items).generate()

    def create_items(self):
        items = []
        for line in self.input:
            quantity, name, price = self.input_parser(line)
            items.append(Item(name, price, quantity))
        return items

    def input_parser(self, line):
        item_parts = line.split()
        quantity = int(item_parts[0])
        name = ' '.join(item_parts[1:-2])
        price = Decimal(item_parts[-1])
        return quantity, name, price

File: lib/receipt_generator.rb
require 'bigdecimal'
require_relative 'receipt'
require_relative 'item'

class ReceiptGenerator
    attr_reader :input

    def initialize(input)
        @input = input
    end

    def get_receipt
        items = create_items
        Receipt.new(items).generate
    end

    def create_items
        items = []
        input.each do |line|
            quantity, name, price = input_parser(line)
            items << Item.new(name, price, quantity)
        end
        items
    end

    def input_parser(line)
        item_parts = line.split(" ")

        quantity = item_parts[0].to_i
        name = item_parts[1..-3].join(" ")
        price = BigDecimal(item_parts.last)

        [quantity, name, price]
    end
end
